mplex/src/lib.rs

```diff
@@ -18,730 +18,372 @@
+{
+}
 
-        let mut outbound: Vec<Substream<_, Buffer>> = vec![];
+impl<C> Write for Substream<C>
+where C: AsyncRead + AsyncWrite
+{
+    fn write(&mut self, buf: &[u8]) -> Result<usize, IoError> {
+        let elem = codec::Elem::Data {
+            substream_id: self.num,
+            data: From::from(buf),
+            endpoint: self.endpoint,
+        };
 
-        for _ in 0..5 {
-            outbound.push(
-                mplex
-                    .clone()
-                    .outbound()
-                    .wait()
-                    .unwrap()
-                    .expect("outbound substream"),
-            );
+        let mut inner = self.inner.lock();
+        match poll_send(&mut inner, elem) {
+            Ok(Async::Ready(())) => Ok(buf.len()),
+            Ok(Async::NotReady) => Err(IoErrorKind::WouldBlock.into()),
+            Err(err) => Err(err),
         }
+    }
 
-        outbound.sort_by_key(|a| a.id());
-
-        for (i, substream) in outbound.iter_mut().enumerate() {
-            assert!(
-                tokio::write_all(substream, i.to_string().as_bytes())
-                    .wait()
-                    .is_ok()
-            );
+    fn flush(&mut self) -> Result<(), IoError> {
+        let mut inner = self.inner.lock();
+        match inner.inner.poll_complete() {
+            Ok(Async::Ready(())) => Ok(()),
+            Ok(Async::NotReady) => Err(IoErrorKind::WouldBlock.into()),
+            Err(err) => Err(err),
         }
+    }
+}
 
-        let stream = io::Cursor::new(mplex.state.lock().wait().unwrap().stream.get_ref().clone());
-
-        let mplex = BufferedMultiplex::<_, Buffer>::listen(stream);
-
-        let mut inbound: Vec<Substream<_, Buffer>> = vec![];
-
-        for _ in 0..5 {
-            let inb: Substream<_, Buffer> = mplex
-                .clone()
-                .inbound()
-                .wait()
-                .unwrap()
-                .expect("inbound substream");
-            inbound.push(inb);
-        }
+impl<C> AsyncWrite for Substream<C>
+where C: AsyncRead + AsyncWrite
+{
+    fn shutdown(&mut self) -> Poll<(), IoError> {
+        let elem = codec::Elem::Close {
+            substream_id: self.num,
+            endpoint: self.endpoint,
+        };
 
-        inbound.sort_by_key(|a| a.id());
-
-        // Skip the first substream and let it be cached.
-        for (mut substream, outbound) in inbound.iter_mut().zip(outbound.iter()).skip(1) {
-            let id = outbound.id();
-            assert_eq!(id, substream.id());
-            assert_eq!(
-                substream
-                    .name()
-                    .and_then(|bytes| String::from_utf8(bytes.to_vec()).ok()),
-                Some(id.to_string())
-            );
-
-            let mut buf = [0; 3];
-            assert_eq!(tokio::read(&mut substream, &mut buf).wait().unwrap().2, 1);
-        }
+        let mut inner = self.inner.lock();
+        poll_send(&mut inner, elem)
+    }
+}
```

Comment:
Since there is no guarantee that `shutdown` is ever called, would it make sense to also remember dropped `Substream`s which have not been shutdown properly and send a close or reset frame eventually? Otherwise subtreams may just be dropped and the remote is never notified.

Reply:
I added a temporary fix for now (`shutdown()` is called in the destructor), but I realized that a proper fix would be way harder.